File: CSC301/HW8/knapsackDC.java
package CSC301.HW8;

import java.util.Arrays;
import java.util.Collections;
import java.io.*;




public class knapsackDC {
    public static void main(String[] args) throws IOException {
     
        System.out.println("See this if it is HW8");
        
        //unit test
        fileReader readerS = new fileReader("CSC301\\HW7\\small.txt");
        fileReader readerM = new fileReader("CSC301\\HW7\\medium.txt");
        fileReader readerL = new fileReader("CSC301\\HW7\\large.txt");
    
         
        int profitS[] = new int[readerS.getValue().size()];
        int weightS[] = new int[readerS.getWeight().size()];
       
        for (int i = 0; i < weightS.length; i++) {
            profitS[i] = readerS.getValue().get(i);
            weightS[i] = readerS.getWeight().get(i);
        }
        
        int[] temp = divide(readerS.W, weightS, profitS);
        int count = 0;
        for (int i = 0; i < temp.length; i++) {
            System.out.print(temp[i] + " ");
            if(temp[i]!= 0)
                count++;
        }
        int V = 0;
        
        int[] re1 = new int[count];
        int t = 0;
        for (int i = 0; i < temp.length-2; i++) {
            if(temp[i]!= 0){
                re1[t] = temp[i];
                System.out.print(re1[t]);
                t++;
            }
        }
        for (int i = 0; i < re1.length; i++) {
            for (int j = profitS.length-1; j >= 0; j--) {
                if(re1[i] == weightS[j]){
                    V += profitS[j];
                    readerS.store.add(j+1);
                }
            }
        }
        Collections.reverse(readerS.store);
        readerS.writeAns(V, "CSC301\\HW8\\DCsmallerout.txt");
        
        
        
        
        

        
        /* 
        int profitM[] = new int[readerM.getValue().size()];
        int weightM[] = new int[readerM.getWeight().size()];
        for (int i = 0; i < weightM.length; i++) {
            profitM[i] = readerM.getValue().get(i);
            weightM[i] = readerM.getWeight().get(i);
        }


        System.out.println(Arrays.toString(divide(readerM.W, weightM, profitM)));
        
        
       
        
        
       
        
        //reading and printing
        
        int profitL[] = new int[readerL.getValue().size()];
        int weightL[] = new int[readerL.getWeight().size()];
        for (int i = 0; i < weightL.length; i++) {
            profitL[i] = readerL.getValue().get(i);
            weightL[i] = readerL.getWeight().get(i);
        }
        System.out.println(divide(readerL.W, weightL, profitL));
        int[] what = divide(readerL.W, weightL, profitL);
        int sum = 0;
        for (int i = 0; i < what.length; i++) {
            if(what[i]!= 0){
                sum += what[i];
            }
        }
        System.out.println("asuchiwopefh: " + sum);
        */
    
    }

    public static int knapsack(int W, int[] w, int[] val, int n){
        int read = 0;
        int write = 1;
        int tmp;
        int readM = 0;
        int writeM = 1;
        int temp;
        int[][] m = new int[2][W+1];
        for (int i = 0; i < W+1; i++) {
            m[0][i] = i;
        }
        /* 
        for (int i = 0; i < m[0].length; i++) {
            System.out.println(m[0][i] +  " " + m[1][i]);
        }
        */


        //memory-efficient
        int[][] bag = new int[2][W+1];
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j <= W; j++) { 
                if(w[i-1] <= j){
                    
                    bag[write][j] = Math.max(bag[read][j],(val[i-1]) + bag[read][j-(w[i-1])]); 
                    if(i > (n+1)/2){
                        if(bag[write][j] == bag[read][j]){
                            m[writeM][j] = m[readM][j];
                            
                        }
                        else{
                            m[writeM][j] = m[readM][j-w[i-1]];
                        }
                        
                    }
                   
                }
                else{
                    bag[write][j] = bag[read][j];
                    if(i > (n+1)/2)
                        m[writeM][j] = m[readM][j];

                }
    
            }
            /*
            
            System.out.println("Outputting!");
            for (int k = 0; k < m[0].length; k++) {
                System.out.println(m[0][k] +  " " + m[1][k]);
            }
            */
            
            if(i > (n+1)/2){
                temp = readM;
                readM = writeM;
                writeM = temp;
            }
            tmp = read;
            read = write;
            write = tmp;     
        }
        
        return m[read][W];

    }
    
    
    public static int[] divide(int W, int[] w, int[] val){
        int k = knapsack(W, w, val, val.length);

        System.out.println(Arrays.toString(w));
        System.out.println("K "+ k + " W " + (W));

        int mid = (w.length-1) / 2;
   
        if(val.length == 2){
            
            int[] ans = new int[2];
            if(k != 0 || (W-k) != 0){
                if((W+k) == w[0] || (W+k) == w[1]){
                    if(w[0] == (W+k))
                        ans[0] = w[0];
                    else
                        ans[0] = w[1];
                    return ans;
                }
                if(w[0] == w[1]){
                    if(val[0] > val[1])
                        ans[0] = w[0];
                    else
                        ans[0] = w[1];
                    return ans;
                }
                if(w[1] == k)
                    ans[0] = w[1];
                else if(w[0] == k)
                    ans[0] = w[0];
                
            }
            return ans;
        }


        int[] wl = new int[mid+1];
        int[] wr = new int[w.length-mid];
        int[] vl = new int[mid+1];
        int[] vr = new int[w.length-mid];
        for (int i = 0; i < mid+1; i++) {
            wl[i] = w[i];
            vl[i] = val[i];
        }
 
        int t = 0;
        for (int i = mid; i < w.length; i++) {
            wr[t] = w[i];
            vr[t] = val[i];
            t++;
        }
        

        //System.out.println(Arrays.toString(wl));
        //System.out.println(Arrays.toString(wr));
         
        int[] num1 = divide(k, wl, vl);
        int[] num2 = divide(W-k, wr, vr);
        int[] sum = new int[num1.length+num2.length];
        
        for (int i = 0; i < num1.length; i++) {
            sum[i] = num1[i];
            
        }
        t = 0;
        for (int i = num1.length; i < sum.length; i++) {
            sum[i] = num2[t];
            t++;
        }
        
        return sum;
            

    }
}
    

    

File: CSC301/HW8/knapsackDC.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <math.h>
#include <string>
#include <fstream>
#include <sstream>

using namespace std;
int knapsack_opt(int W, vector<int> w, vector<int> val, int n){
    /*this method is essentially the same as knapasck_cap but
    with different returning value so I just copy it.
    */
    int read = 0;
    int write = 1;
    int tmp;
    int readM = 0;
    int writeM = 1;
    int temp;
    int m[2][W+1];

    //intialization
    for (int i = 0; i < W + 1; i++) {
            m[0][i] = i;
    }
    
    int bag[2][W+1];
    for (int i = 0; i < W + 1; i++) {
            bag[0][i] = 0;
            bag[1][i] = 0;
    }
    
    //tracing with two arrays
    for (int i = 1; i <= n; i++) {
            for (int j = 0; j <= W; j++) {
                if (w[i - 1] <= j) {
                    bag[write][j] = max(bag[read][j], (val[i - 1] + bag[read][j - (w[i - 1])]));
                    if (i > (n + 1) / 2) {//start updating m
                        if (bag[write][j] == bag[read][j]) {
                            m[writeM][j] = m[readM][j];
                        } else 
                            m[writeM][j] = m[readM][j - w[i - 1]];
                        
                    }

                } else {
                    bag[write][j] = bag[read][j];
                    if (i > (n + 1) / 2)
                        m[writeM][j] = m[readM][j];

                }

            }
            //updating prev and curr pointer
            if (i > (n + 1) / 2) {
                temp = readM;
                readM = writeM;
                writeM = temp;
            }

            tmp = read;
            read = write;
            write = tmp;
        }

        return bag[read][W];

}
int knapsack_cap(int W, vector<int> w, vector<int> val, int n){
    int read = 0;
    int write = 1;
    int tmp;
    int readM = 0;
    int writeM = 1;
    int temp;
    int m[2][W+1];

    //intialization
    for (int i = 0; i < W + 1; i++) {
            m[0][i] = i;
    }
    
    int bag[2][W+1];
    for (int i = 0; i < W + 1; i++) {
            bag[0][i] = 0;
            bag[1][i] = 0;
    }
    
    //tracing with two arrays
    for (int i = 1; i <= n; i++) {
            for (int j = 0; j <= W; j++) {
                if (w[i - 1] <= j) {
                    //normal updating with bag
                    bag[write][j] = max(bag[read][j], (val[i - 1] + bag[read][j - (w[i - 1])]));
                    if (i > (n + 1) / 2) {//start updating m
                        if (bag[write][j] == bag[read][j]) {
                            m[writeM][j] = m[readM][j];

                        } else 
                            m[writeM][j] = m[readM][j - w[i - 1]];
                    }

                } else {
                    bag[write][j] = bag[read][j];
                    if (i > (n + 1) / 2)
                        m[writeM][j] = m[readM][j];
                }
            }
            //updating prev and curr pointer
            if (i > (n + 1) / 2) {
                temp = readM;
                readM = writeM;
                writeM = temp;
            }
            tmp = read;
            read = write;
            write = tmp;
        }

        return m[read][W];

}

vector<pair<int, int>> last(vector<int> val, vector<int> w, int W) {
    //creating one-to-one correspondence of current val and elements.
    vector<pair<int, int>> K(W+1, {0,-1});
    
    for (int i = 0; i < val.size(); ++i) {
        for (int j = W; j >= 0; --j) {
            if (w[i] <= j){
                if(K[j].first < K[j - w[i]].first + val[i]) 
                    K[j] = { K[j - w[i]].first + val[i], i };
            }
        }
    }
    return K;
}

vector<int> knapsack_dc(vector<int>  v, vector<int>  w, int W, int offset = 0) {
    //base case
    if (v.empty()) {
        return {};
    }
    //middle capacity
    int k = knapsack_cap(W, w, v, w.size());
    
    //partitioning 
    int mid = v.size() / 2;
    
    vector<pair<int, int>> subSol1 = last(vector<int>(begin(v), begin(v) + mid), vector<int>(begin(w), begin(w) + mid), W);
    vector<pair<int, int>> subSol2 = last(vector<int>(begin(v) + mid, end(v)), vector<int>(begin(w) + mid, end(w)), W);
    

    pair<int, int> best = { -1, -1 };
    for (int i = 0; i <= W; ++i) {
        //maximum the value and the capacity
        best = max(best, { subSol1[i].first + subSol2[W - i].first, i });
    }
    if(k != best.second)
        k = best.second;
    cout << best.first << " " << best.second << endl;

    vector<int> solution;   
    if (subSol1[k].second != -1) {
        
        //get the chosen item
        int iChosen = subSol1[best.second].second;

        //continue recursion after delete capacity after choosing 
        solution = knapsack_dc(vector<int>(begin(v), begin(v) + iChosen), vector<int>(begin(w), begin(w) + iChosen), k - w[iChosen], offset);
        solution.push_back(subSol1[best.second].second + offset+1);
    }
    if (subSol2[W - k].second != -1) {
        
        //get the chosen item
        int iChosen = mid + subSol2[W - best.second].second;
        //continue recursion after delete capacity after choosing 
        auto subSolution = knapsack_dc(vector<int>(begin(v) + mid, begin(v) + iChosen), vector<int>(begin(w) + mid, begin(w) + iChosen), W - k - w[iChosen], offset + mid);
        //merge two arrays
        copy(begin(subSolution), end(subSolution), back_inserter(solution));
        //pushing index into sol
        solution.push_back(iChosen + offset+1);
    }

    return solution;
}

int main(){
    //small test
    vector<int> vect;
    
    ifstream small("small.txt");
    string line;
    getline(small,line);
    cout << line.substr(2,line.size()) << endl;
    
    int SW = stoi(line.substr(2,line.size()), NULL, 10);
    getline(small, line);
    cout << line.substr(2,line.size()) << endl;
    int Sn = stoi(line.substr(2,line.size()), NULL, 10);
    vector<int> smallW;
    vector<int> smallV;

    while(small.good()){
        int count = 0;
        stringstream ss;
        string temp;
        getline(small, line);
        ss << line;
        int num;
        while(!ss.eof()){
            ss >> temp;
            if(stringstream(temp) >> num){
                if(count == 0){
                    smallW.push_back(num);
                    

                }
                else{
                    smallV.push_back(num);

                }
                count++;
            }
        }
    }
    
    int test2 = knapsack_opt(SW, smallW, smallV, Sn);
    vect = knapsack_dc(smallV,smallW, SW, 0);
    
   
    cout << "how " << test2 << endl;

    FILE * smallOut;
    smallOut = fopen("smallDCout.txt", "w+");
    fprintf(smallOut, "V %d\n", test2);
    fprintf(smallOut, "i %d\n", vect.size());
    for(int i = 0; i < vect.size(); i++){
        fprintf(smallOut, "%d\n", vect[i]);

    }

    fclose(smallOut);
    vect.clear();


    //medium test
    ifstream medium("medium.txt");
    string line2;
    getline(medium,line2);
    cout << line2.substr(2,line2.size()) << endl;
    
    int MW = stoi(line2.substr(2,line2.size()), NULL, 10);
    getline(medium, line2);
    cout << line2.substr(2,line2.size()) << endl;
    int Mn = stoi(line2.substr(2,line2.size()), NULL, 10);
    vector<int> mediumW;
    vector<int> mediumV;

    while(medium.good()){
        int count = 0;
        stringstream ss;
        string temp;
        getline(medium, line2);
        ss << line2;
        int num;
        while(!ss.eof()){
            ss >> temp;
            if(stringstream(temp) >> num){
                if(count == 0){
                    mediumW.push_back(num);
                    

                }
                else{
                    mediumV.push_back(num);

                }
                count++;
            }
        }
    }

    int test3 = knapsack_opt(MW, mediumW, mediumV, Mn);

    vect = knapsack_dc(mediumV, mediumW, MW, 0);


    FILE * mediumOut;
    mediumOut = fopen("mediumDCout.txt", "w+");
    fprintf(mediumOut, "V %d\n", test3);
    fprintf(mediumOut, "i %d\n", vect.size());
    for(int i = 0; i < vect.size(); i++){
        fprintf(mediumOut, "%d\n", vect[i]);

    }

    fclose(mediumOut);
    vect.clear();
   
    
    cout << "how " << test3 << endl;

    //large test
    ifstream large("large.txt");
    string line3;
    getline(large,line3);
    cout << line3.substr(2,line3.size()) << endl;
    
    int LW = stoi(line3.substr(2,line3.size()), NULL, 10);
    getline(large, line3);
    cout << line3.substr(2,line3.size()) << endl;
    int Ln = stoi(line3.substr(2,line3.size()), NULL, 10);
    vector<int> largeW;
    vector<int> largeV;

    while(large.good()){
        int count = 0;
        stringstream ss;
        string temp;
        getline(large, line3);
        ss << line3;
        int num;
        while(!ss.eof()){
            ss >> temp;
            if(stringstream(temp) >> num){
                if(count == 0){
                    largeW.push_back(num);
                }
                else{
                    largeV.push_back(num);
                }
                count++;
            }
        }
    }
    int test4 = knapsack_opt(LW, largeW, largeV, Ln);
    vect = knapsack_dc(largeV, largeW, LW, 0);



    FILE * largeOut;
    largeOut = fopen("largeDCout.txt", "w+");
    fprintf(largeOut, "V %d\n", test4);
    fprintf(largeOut, "i %d\n", vect.size());
    for(int i = 0; i < vect.size(); i++){
        fprintf(largeOut, "%d\n", vect[i]);

    }

    fclose(largeOut);


    vect.clear();
   
    
    cout << "how " << test4 << endl;
    
}